File: rb/GithubProjectIssues_sdk.rb
# GithubProjectIssues SDK

require_relative 'utility/struct/voxgig_struct'
require_relative 'core/utility_type'
require_relative 'core/spec'
require_relative 'core/helpers'

# Load utility registration
require_relative 'utility/register'

# Load config and features
require_relative 'config'
require_relative 'feature/base_feature'
require_relative 'features'

# Load typed models (Struct value objects).
require_relative 'GithubProjectIssues_types'


class GithubProjectIssuesSDK
  attr_accessor :mode, :features, :options

  def initialize(options = {})
    @mode = "live"
    @features = []
    @options = nil

    utility = GithubProjectIssuesUtility.new
    @_utility = utility

    config = GithubProjectIssuesConfig.make_config

    @_rootctx = utility.make_context.call({
      "client" => self,
      "utility" => utility,
      "config" => config,
      "options" => options || {},
      "shared" => {},
    }, nil)

    @options = utility.make_options.call(@_rootctx)

    if VoxgigStruct.getpath(@options, "feature.test.active") == true
      @mode = "test"
    end

    @_rootctx.options = @options

    # Add features from config.
    feature_opts = GithubProjectIssuesHelpers.to_map(VoxgigStruct.getprop(@options, "feature"))
    if feature_opts
      items = VoxgigStruct.items(feature_opts)
      if items
        items.each do |item|
          fname = item[0]
          fopts = GithubProjectIssuesHelpers.to_map(item[1])
          if fopts && fopts["active"] == true
            utility.feature_add.call(@_rootctx, GithubProjectIssuesFeatures.make_feature(fname))
          end
        end
      end
    end

    # Add extension features.
    extend_val = VoxgigStruct.getprop(@options, "extend")
    if extend_val.is_a?(Array)
      extend_val.each do |f|
        if f.respond_to?(:get_name)
          utility.feature_add.call(@_rootctx, f)
        end
      end
    end

    # Initialize features.
    @features.each do |f|
      utility.feature_init.call(@_rootctx, f)
    end

    utility.feature_hook.call(@_rootctx, "PostConstruct")
  end

  def options_map
    out = VoxgigStruct.clone(@options)
    out.is_a?(Hash) ? out : {}
  end

  def get_utility
    GithubProjectIssuesUtility.copy(@_utility)
  end

  def get_root_ctx
    @_rootctx
  end

  def prepare(fetchargs = {})
    utility = @_utility
    fetchargs ||= {}

    ctrl = GithubProjectIssuesHelpers.to_map(VoxgigStruct.getprop(fetchargs, "ctrl")) || {}

    ctx = utility.make_context.call({
      "opname" => "prepare",
      "ctrl" => ctrl,
    }, @_rootctx)

    opts = @options
    path = VoxgigStruct.getprop(fetchargs, "path") || ""
    path = "" unless path.is_a?(String)
    method_val = VoxgigStruct.getprop(fetchargs, "method") || "GET"
    method_val = "GET" unless method_val.is_a?(String)
    params = GithubProjectIssuesHelpers.to_map(VoxgigStruct.getprop(fetchargs, "params")) || {}
    query = GithubProjectIssuesHelpers.to_map(VoxgigStruct.getprop(fetchargs, "query")) || {}
    headers = utility.prepare_headers.call(ctx)

    base = VoxgigStruct.getprop(opts, "base") || ""
    base = "" unless base.is_a?(String)
    prefix = VoxgigStruct.getprop(opts, "prefix") || ""
    prefix = "" unless prefix.is_a?(String)
    suffix = VoxgigStruct.getprop(opts, "suffix") || ""
    suffix = "" unless suffix.is_a?(String)

    ctx.spec = GithubProjectIssuesSpec.new({
      "base" => base, "prefix" => prefix, "suffix" => suffix,
      "path" => path, "method" => method_val,
      "params" => params, "query" => query, "headers" => headers,
      "body" => VoxgigStruct.getprop(fetchargs, "body"),
      "step" => "start",
    })

    # Merge user-provided headers.
    uh = VoxgigStruct.getprop(fetchargs, "headers")
    if uh.is_a?(Hash)
      uh.each { |k, v| ctx.spec.headers[k] = v }
    end

    _, err = utility.prepare_auth.call(ctx)
    raise err if err

    utility.make_fetch_def.call(ctx)
  end

  def direct(fetchargs = {})
    utility = @_utility

    # direct() is the raw-HTTP escape hatch: it always returns a result hash
    # ({ "ok" => ..., ... }) and never raises. prepare() raises on error, so
    # trap that and surface it in the hash.
    begin
      fetchdef = prepare(fetchargs)
    rescue GithubProjectIssuesError => err
      return { "ok" => false, "err" => err }
    end

    fetchargs ||= {}
    ctrl = GithubProjectIssuesHelpers.to_map(VoxgigStruct.getprop(fetchargs, "ctrl")) || {}

    ctx = utility.make_context.call({
      "opname" => "direct",
      "ctrl" => ctrl,
    }, @_rootctx)

    url = fetchdef["url"] || ""
    fetched, fetch_err = utility.fetcher.call(ctx, url, fetchdef)

    return { "ok" => false, "err" => fetch_err } if fetch_err

    if fetched.nil?
      return {
        "ok" => false,
        "err" => ctx.make_error("direct_no_response", "response: undefined"),
      }
    end

    if fetched.is_a?(Hash)
      status = GithubProjectIssuesHelpers.to_int(VoxgigStruct.getprop(fetched, "status"))
      headers = VoxgigStruct.getprop(fetched, "headers") || {}

      # No-body responses (204, 304) and explicit zero content-length must
      # skip JSON parsing — calling json() on an empty body errors.
      content_length = headers.is_a?(Hash) ? headers["content-length"] : nil
      no_body = status == 204 || status == 304 || content_length.to_s == "0"

      json_data = nil
      unless no_body
        jf = VoxgigStruct.getprop(fetched, "json")
        if jf.is_a?(Proc)
          begin
            json_data = jf.call
          rescue StandardError
            # Non-JSON body — leave data nil, keep status/headers.
            json_data = nil
          end
        end
      end

      return {
        "ok" => status >= 200 && status < 300,
        "status" => status,
        "headers" => headers,
        "data" => json_data,
      }
    end

    return {
      "ok" => false,
      "err" => ctx.make_error("direct_invalid", "invalid response type"),
    }
  end


  # Idiomatic facade: client.coffee.list / client.coffee.load({ "id" => ... })
  def coffee
    require_relative 'entity/coffee_entity'
    @coffee ||= CoffeeEntity.new(self, nil)
  end

  # Deprecated: use client.coffee instead.
  def Coffee(data = nil)
    require_relative 'entity/coffee_entity'
    CoffeeEntity.new(self, data)
  end


  # Idiomatic facade: client.coffee_domain.list / client.coffee_domain.load({ "id" => ... })
  def coffee_domain
    require_relative 'entity/coffee_domain_entity'
    @coffee_domain ||= CoffeeDomainEntity.new(self, nil)
  end

  # Deprecated: use client.coffee_domain instead.
  def CoffeeDomain(data = nil)
    require_relative 'entity/coffee_domain_entity'
    CoffeeDomainEntity.new(self, data)
  end


  # Idiomatic facade: client.donate_rest_controller.list / client.donate_rest_controller.load({ "id" => ... })
  def donate_rest_controller
    require_relative 'entity/donate_rest_controller_entity'
    @donate_rest_controller ||= DonateRestControllerEntity.new(self, nil)
  end

  # Deprecated: use client.donate_rest_controller instead.
  def DonateRestController(data = nil)
    require_relative 'entity/donate_rest_controller_entity'
    DonateRestControllerEntity.new(self, data)
  end


  # Idiomatic facade: client.portfolio_controller.list / client.portfolio_controller.load({ "id" => ... })
  def portfolio_controller
    require_relative 'entity/portfolio_controller_entity'
    @portfolio_controller ||= PortfolioControllerEntity.new(self, nil)
  end

  # Deprecated: use client.portfolio_controller instead.
  def PortfolioController(data = nil)
    require_relative 'entity/portfolio_controller_entity'
    PortfolioControllerEntity.new(self, data)
  end


  # Idiomatic facade: client.repository_detail_domain.list / client.repository_detail_domain.load({ "id" => ... })
  def repository_detail_domain
    require_relative 'entity/repository_detail_domain_entity'
    @repository_detail_domain ||= RepositoryDetailDomainEntity.new(self, nil)
  end

  # Deprecated: use client.repository_detail_domain instead.
  def RepositoryDetailDomain(data = nil)
    require_relative 'entity/repository_detail_domain_entity'
    RepositoryDetailDomainEntity.new(self, data)
  end


  # Idiomatic facade: client.repository_issue_domain.list / client.repository_issue_domain.load({ "id" => ... })
  def repository_issue_domain
    require_relative 'entity/repository_issue_domain_entity'
    @repository_issue_domain ||= RepositoryIssueDomainEntity.new(self, nil)
  end

  # Deprecated: use client.repository_issue_domain instead.
  def RepositoryIssueDomain(data = nil)
    require_relative 'entity/repository_issue_domain_entity'
    RepositoryIssueDomainEntity.new(self, data)
  end


  # Idiomatic facade: client.version.list / client.version.load({ "id" => ... })
  def version
    require_relative 'entity/version_entity'
    @version ||= VersionEntity.new(self, nil)
  end

  # Deprecated: use client.version instead.
  def Version(data = nil)
    require_relative 'entity/version_entity'
    VersionEntity.new(self, data)
  end



  def self.test(testopts = nil, sdkopts = nil)
    sdkopts = sdkopts || {}
    sdkopts = VoxgigStruct.clone(sdkopts)
    sdkopts = {} unless sdkopts.is_a?(Hash)

    testopts = testopts || {}
    testopts = VoxgigStruct.clone(testopts)
    testopts = {} unless testopts.is_a?(Hash)
    testopts["active"] = true

    VoxgigStruct.setpath(sdkopts, "feature.test", testopts)

    sdk = GithubProjectIssuesSDK.new(sdkopts)
    sdk.mode = "test"
    sdk
  end
end

File: py/githubprojectissues_sdk.py
# GithubProjectIssues SDK

from utility.voxgig_struct import voxgig_struct as vs
from core.utility_type import GithubProjectIssuesUtility
from core.spec import GithubProjectIssuesSpec
from core import helpers

# Load utility registration (populates Utility._registrar)
from utility import register

# Load features
from feature.base_feature import GithubProjectIssuesBaseFeature
from features import _make_feature


class GithubProjectIssuesSDK:

    def __init__(self, options=None):
        self.mode = "live"
        self.features = []
        self.options = None

        utility = GithubProjectIssuesUtility()
        self._utility = utility

        from config import make_config
        config = make_config()

        self._rootctx = utility.make_context({
            "client": self,
            "utility": utility,
            "config": config,
            "options": options if options is not None else {},
            "shared": {},
        }, None)

        self.options = utility.make_options(self._rootctx)

        if vs.getpath(self.options, "feature.test.active") is True:
            self.mode = "test"

        self._rootctx.options = self.options

        # Add features from config.
        feature_opts = helpers.to_map(vs.getprop(self.options, "feature"))
        if feature_opts is not None:
            feature_items = vs.items(feature_opts)
            if feature_items is not None:
                for item in feature_items:
                    fname = item[0]
                    fopts = helpers.to_map(item[1])
                    if fopts is not None and fopts.get("active") is True:
                        utility.feature_add(self._rootctx, _make_feature(fname))

        # Add extension features.
        extend = vs.getprop(self.options, "extend")
        if isinstance(extend, list):
            for f in extend:
                if isinstance(f, dict) or (hasattr(f, "get_name") and callable(f.get_name)):
                    utility.feature_add(self._rootctx, f)

        # Initialize features.
        for f in self.features:
            utility.feature_init(self._rootctx, f)

        utility.feature_hook(self._rootctx, "PostConstruct")

        # #BuildFeatures

    def options_map(self):
        out = vs.clone(self.options)
        if isinstance(out, dict):
            return out
        return {}

    def get_utility(self):
        return GithubProjectIssuesUtility.copy(self._utility)

    def get_root_ctx(self):
        return self._rootctx

    def prepare(self, fetchargs=None):
        utility = self._utility

        if fetchargs is None:
            fetchargs = {}

        ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl"))
        if ctrl is None:
            ctrl = {}

        ctx = utility.make_context({
            "opname": "prepare",
            "ctrl": ctrl,
        }, self._rootctx)

        options = self.options

        path = vs.getprop(fetchargs, "path") or ""
        if not isinstance(path, str):
            path = ""

        method = vs.getprop(fetchargs, "method") or "GET"
        if not isinstance(method, str):
            method = "GET"

        params = helpers.to_map(vs.getprop(fetchargs, "params"))
        if params is None:
            params = {}
        query = helpers.to_map(vs.getprop(fetchargs, "query"))
        if query is None:
            query = {}

        headers = utility.prepare_headers(ctx)

        base = vs.getprop(options, "base") or ""
        if not isinstance(base, str):
            base = ""
        prefix = vs.getprop(options, "prefix") or ""
        if not isinstance(prefix, str):
            prefix = ""
        suffix = vs.getprop(options, "suffix") or ""
        if not isinstance(suffix, str):
            suffix = ""

        ctx.spec = GithubProjectIssuesSpec({
            "base": base,
            "prefix": prefix,
            "suffix": suffix,
            "path": path,
            "method": method,
            "params": params,
            "query": query,
            "headers": headers,
            "body": vs.getprop(fetchargs, "body"),
            "step": "start",
        })

        # Merge user-provided headers.
        uh = vs.getprop(fetchargs, "headers")
        if isinstance(uh, dict):
            for k, v in uh.items():
                ctx.spec.headers[k] = v

        _, err = utility.prepare_auth(ctx)
        if err is not None:
            raise err

        fetchdef, err = utility.make_fetch_def(ctx)
        if err is not None:
            raise err

        return fetchdef

    def direct(self, fetchargs=None):
        utility = self._utility

        try:
            fetchdef = self.prepare(fetchargs)
        except Exception as err:
            # direct() is the raw-HTTP escape hatch: it never raises, it
            # returns a result object callers branch on via result["ok"].
            return {"ok": False, "err": err}

        if fetchargs is None:
            fetchargs = {}
        ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl"))
        if ctrl is None:
            ctrl = {}

        ctx = utility.make_context({
            "opname": "direct",
            "ctrl": ctrl,
        }, self._rootctx)

        url = fetchdef.get("url", "")
        fetched, fetch_err = utility.fetcher(ctx, url, fetchdef)

        if fetch_err is not None:
            return {"ok": False, "err": fetch_err}

        if fetched is None:
            return {
                "ok": False,
                "err": ctx.make_error("direct_no_response", "response: undefined"),
            }

        if isinstance(fetched, dict):
            status = helpers.to_int(vs.getprop(fetched, "status"))
            headers = vs.getprop(fetched, "headers") or {}

            # No-body responses (204, 304) and explicit zero content-length
            # must skip JSON parsing — calling json() on an empty body raises.
            content_length = None
            if isinstance(headers, dict):
                content_length = headers.get("content-length")
            no_body = status in (204, 304) or str(content_length) == "0"

            json_data = None
            if not no_body:
                jf = vs.getprop(fetched, "json")
                if callable(jf):
                    try:
                        json_data = jf()
                    except Exception:
                        # Non-JSON body (e.g. text/plain, text/html). Surface
                        # status + headers but leave data as None.
                        json_data = None

            return {
                "ok": status >= 200 and status < 300,
                "status": status,
                "headers": headers,
                "data": json_data,
            }

        return {
            "ok": False,
            "err": ctx.make_error("direct_invalid", "invalid response type"),
        }


    @property
    def coffee(self):
        """Idiomatic facade: client.coffee.list() / client.coffee.load({"id": ...})."""
        from entity.coffee_entity import CoffeeEntity
        cached = getattr(self, "_coffee", None)
        if cached is None:
            cached = CoffeeEntity(self, None)
            self._coffee = cached
        return cached

    def Coffee(self, data=None):
        # Deprecated: use client.coffee instead.
        from entity.coffee_entity import CoffeeEntity
        return CoffeeEntity(self, data)


    @property
    def coffee_domain(self):
        """Idiomatic facade: client.coffee_domain.list() / client.coffee_domain.load({"id": ...})."""
        from entity.coffee_domain_entity import CoffeeDomainEntity
        cached = getattr(self, "_coffee_domain", None)
        if cached is None:
            cached = CoffeeDomainEntity(self, None)
            self._coffee_domain = cached
        return cached

    def CoffeeDomain(self, data=None):
        # Deprecated: use client.coffee_domain instead.
        from entity.coffee_domain_entity import CoffeeDomainEntity
        return CoffeeDomainEntity(self, data)


    @property
    def donate_rest_controller(self):
        """Idiomatic facade: client.donate_rest_controller.list() / client.donate_rest_controller.load({"id": ...})."""
        from entity.donate_rest_controller_entity import DonateRestControllerEntity
        cached = getattr(self, "_donate_rest_controller", None)
        if cached is None:
            cached = DonateRestControllerEntity(self, None)
            self._donate_rest_controller = cached
        return cached

    def DonateRestController(self, data=None):
        # Deprecated: use client.donate_rest_controller instead.
        from entity.donate_rest_controller_entity import DonateRestControllerEntity
        return DonateRestControllerEntity(self, data)


    @property
    def portfolio_controller(self):
        """Idiomatic facade: client.portfolio_controller.list() / client.portfolio_controller.load({"id": ...})."""
        from entity.portfolio_controller_entity import PortfolioControllerEntity
        cached = getattr(self, "_portfolio_controller", None)
        if cached is None:
            cached = PortfolioControllerEntity(self, None)
            self._portfolio_controller = cached
        return cached

    def PortfolioController(self, data=None):
        # Deprecated: use client.portfolio_controller instead.
        from entity.portfolio_controller_entity import PortfolioControllerEntity
        return PortfolioControllerEntity(self, data)


    @property
    def repository_detail_domain(self):
        """Idiomatic facade: client.repository_detail_domain.list() / client.repository_detail_domain.load({"id": ...})."""
        from entity.repository_detail_domain_entity import RepositoryDetailDomainEntity
        cached = getattr(self, "_repository_detail_domain", None)
        if cached is None:
            cached = RepositoryDetailDomainEntity(self, None)
            self._repository_detail_domain = cached
        return cached

    def RepositoryDetailDomain(self, data=None):
        # Deprecated: use client.repository_detail_domain instead.
        from entity.repository_detail_domain_entity import RepositoryDetailDomainEntity
        return RepositoryDetailDomainEntity(self, data)


    @property
    def repository_issue_domain(self):
        """Idiomatic facade: client.repository_issue_domain.list() / client.repository_issue_domain.load({"id": ...})."""
        from entity.repository_issue_domain_entity import RepositoryIssueDomainEntity
        cached = getattr(self, "_repository_issue_domain", None)
        if cached is None:
            cached = RepositoryIssueDomainEntity(self, None)
            self._repository_issue_domain = cached
        return cached

    def RepositoryIssueDomain(self, data=None):
        # Deprecated: use client.repository_issue_domain instead.
        from entity.repository_issue_domain_entity import RepositoryIssueDomainEntity
        return RepositoryIssueDomainEntity(self, data)


    @property
    def version(self):
        """Idiomatic facade: client.version.list() / client.version.load({"id": ...})."""
        from entity.version_entity import VersionEntity
        cached = getattr(self, "_version", None)
        if cached is None:
            cached = VersionEntity(self, None)
            self._version = cached
        return cached

    def Version(self, data=None):
        # Deprecated: use client.version instead.
        from entity.version_entity import VersionEntity
        return VersionEntity(self, data)



    @classmethod
    def test(cls, testopts=None, sdkopts=None):
        if sdkopts is None:
            sdkopts = {}
        sdkopts = vs.clone(sdkopts)
        if not isinstance(sdkopts, dict):
            sdkopts = {}

        if testopts is None:
            testopts = {}
        testopts = vs.clone(testopts)
        if not isinstance(testopts, dict):
            testopts = {}
        testopts["active"] = True

        vs.setpath(sdkopts, "feature.test", testopts)

        sdk = cls(sdkopts)
        sdk.mode = "test"

        return sdk
